src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -108,58 +122,95 @@ public PaymentHandleResponse patchAddress(@Nonnull Cart cart, @Nonnull String pa
+        String updateTimeStr = paypalPayment.getUpdateTime() == null ? paypalPayment.getCreateTime() : paypalPayment.getUpdateTime();
+        TransactionDraft transactionDraft = TransactionDraftBuilder
+                .of(TransactionType.CHARGE, Money.of(total, totalAmount.getCurrency()))
+                .timestamp(toZonedDateTime(updateTimeStr))
+                .state(transactionState)
+                .build();
+        return ctpFacade.getPaymentService()
+                .updatePayment(ctpPaymentId, Collections.singletonList(AddTransaction.of(transactionDraft)))
+                .toCompletableFuture()
+                .join();
+    }
+
+    protected CompletionStage<Payment> setApprovalUrlAndInterfaceId(Payment newPpPayment, String ctpPaymentId) {
         List<UpdateAction<io.sphere.sdk.payments.Payment>> updateActions = asList(
                 SetCustomField.ofObject(APPROVAL_URL, PaymentMapper.getApprovalUrl(newPpPayment).orElse("")),
                 SetInterfaceId.of(newPpPayment.getId()));
         return ctpFacade.getPaymentService().updatePayment(ctpPaymentId, updateActions)
                 .thenApply(ignore -> newPpPayment);
     }
+
+    private CompletionStage<PaymentHandleResponse> executePaymentAndCreateTxn(@Nonnull String paypalPlusPaymentId,
+                                                                              @Nonnull String paypalPlusPayerId,
+                                                                              @Nonnull io.sphere.sdk.payments.Payment ctpPayment) {
+        return paypalPlusFacade.getPaymentService().execute(new Payment().setId(paypalPlusPaymentId),
+                new PaymentExecution().setPayerId(paypalPlusPayerId))
+                .thenApply(paypalPayment -> {
+                    if (PaypalPlusPaymentStates.APPROVED.equals(paypalPayment.getState())) {
+                        return createChargeTransaction(paypalPayment, ctpPayment.getId(), TransactionState.SUCCESS);
+                    } else {
+                        throw new PaypalPlusPaymentException(format("Error when approving payment [%s], current state=[%s]",
+                                paypalPlusPaymentId, paypalPayment.getState()));
+                    }
+                })
+                .thenApply(payment -> PaymentHandleResponse.ofStatusCode(HttpStatus.OK));
+    }
+
+    private PaymentHandleResponse paymentExceptionWrapper(String paypalPlusPaymentId,
+                                                          CompletionStage<PaymentHandleResponse> completionStage) {
+        return completionStage
+                .exceptionally(throwable -> {
+                    logger.error("Unexpected exception processing Paypal Plus paymentId=[{}]:", paypalPlusPaymentId, throwable);
+                    return PaymentHandleResponse.of400BadRequest(
+                            format("PaypalPlus payment [%s] can't be processed, see the logs", paypalPlusPaymentId));
+                }).toCompletableFuture().join();
```

Comment:
I would block where it is called and not here. This method should stay async so that we do not need to touch it on refactor. Please avoid usage of `.join` otherwise you risk to block the thread forever. In order to omit it:

- use `.get(timeout, unit)` or proper CTP blocking client (see also area with bad examples: http://commercetools.github.io/commercetools-jvm-sdk/apidocs/io/sphere/sdk/meta/SphereClientTuningDocumentation.html#blocking-client)
- for PayPal calls use: `.get(timeout, unit)`